### PR TITLE
[WIP] add cyclotomic square of fp12 element

### PR DESCRIPTION
**Pull request by Raneet10**
# Add cyclotomic square of `fp12` element

## Description

This PR attempts to add [faster squaring for 6th cyclotomic subgroup elements](https://eprint.iacr.org/2009/565.pdf). 
I imagine this would be the first step towards implementing #544 

I tried implementing what I described in [this comment ](https://github.com/lambdaclass/lambdaworks/issues/544#issuecomment-1925423017) but there are obvious gaps in my knowledge which is apparent in the code. Hence, I am hoping to get reviews here to bridge those gaps and clear away my misunderstandings.   

## Type of change

Please delete options that are not relevant.

- [x] New feature
- [ ] Bug fix
- [x] Optimization

## Checklist
- [x] Linked to Github Issue
- [x] Unit tests added
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.
- [x] This change is an Optimization
  - [ ] Benchmarks added/run

### CODE REVIEW

File: math/src/elliptic_curve/short_weierstrass/curves/bls12_381/field_extension.rs [hunk 1]
```diff
@@ -282,6 +283,102 @@ impl FieldElement<Degree12ExtensionField> {
             ]),
         ])
     }
+
+    // Follows https://www.iacr.org/archive/pkc2010/60560212/60560212.pdf section 3.2
+    pub fn cyclotomic_square(&self) -> Self {
+        // f² = A' + B'⍵² + 2⍵*√A * √B , where A' = A₀ + A₁y + A₂y² , B' = B₀ + B₁y + B₂y² and x², y = ⍵²
```
Comment (mratsim) [inline math/src/elliptic_curve/short_weierstrass/curves/bls12_381/field_extension.rs:289]: You need to document clearly whether you use quadratic over cubic (Fp2->Fp6->Fp12) or cubic over quadratic like in the paper (Fp2->Fp4->Fp12)

**Review by mratsim — COMMENTED** [review]
As mentioned in Telegram

It's important to mention in document that the paper uses a Fp2->Fp4->Fp12 towering while usually people implement Fp2->Fp6->Fp12.

Formulas need to be adapted like this depending on towering: https://github.com/mratsim/constantine/blob/master/constantine%2Fmath%2Fpairings%2Fcyclotomic_subgroups.nim#L278

File: math/src/elliptic_curve/short_weierstrass/curves/bls12_381/field_extension.rs (hunk 1 above)
Comment (Raneet10) [reply math/src/elliptic_curve/short_weierstrass/curves/bls12_381/field_extension.rs:289]: > You need to document clearly whether you use quadratic over cubic (Fp2->Fp6->Fp12) or cubic over quadratic like in the paper (Fp2->Fp4->Fp12)

Yeah that's a good point. Though I felt mentioning the section in the paper would be self explanatory.

**Comment by Raneet10** [comment]
> As mentioned in Telegram
> 
> It's important to mention in document that the paper uses a Fp2->Fp4->Fp12 towering while usually people implement Fp2->Fp6->Fp12.
> 
> Formulas need to be adapted like this depending on towering: https://github.com/mratsim/constantine/blob/master/constantine%2Fmath%2Fpairings%2Fcyclotomic_subgroups.nim#L278

Thanks for the reference! 
Unless I'm missing something, the paper seems to derive the squaring for `Fp6` (by cubic over quadratic, `Fp -> Fp2 -> Fp6`) and not `Fp12`. I tried to follow the same train of  thought [here in this rundown ](https://github.com/lambdaclass/lambdaworks/issues/544#issuecomment-1925423017), which I believe is quadratic over cubic over quadratic (`Fp -> Fp2 -> Fp6 -> Fp12`). Could you maybe go through that and let me know if it makes sense ?

**Comment by mratsim** [comment]
> Unless I'm missing something, the paper seems to derive the squaring for Fp6 (by cubic over quadratic, Fp -> Fp2 -> Fp6) and not Fp12. I tried to follow the same train of thought https://github.com/lambdaclass/lambdaworks/issues/544#issuecomment-1925423017, which I believe is quadratic over cubic over quadratic (Fp -> Fp2 -> Fp6 -> Fp12). Could you maybe go through that and let me know if it makes sense ?

The paper is about cyclotomic subgroups first and foremost, and use Fp6 as an example with Fp->Fp2->Fp6. You're using Fp12, so paper formulas are for Fp2->Fp4->Fp12.

Hence either you match the paper cubic over quadratic or you adapt them.

**Comment by mratsim** [comment]
> So should I be deriving the formula for `Fp6` via `Fp -> Fp3 -> Fp6`

No, you forgot about `Fp`, completely. What matters is how Fp12 is built as a 6th degree extension. And it's built as a 6th degree extension from Fp2.

**Comment by Raneet10** [comment]
> > So should I be deriving the formula for `Fp6` via `Fp -> Fp3 -> Fp6`
> 
> No, you forgot about `Fp`, completely. What matters is how Fp12 is built as a 6th degree extension. And it's built as a 6th degree extension from Fp2.

Okay I see your point. I'll try to revaluate.

**Comment by diegokingston** [comment]
This functionality has been added

**Pull request closed without merging**